강원대 BE_이시행 3단계 - 상품 옵션

sweet-mine opened the pull request:
options 테이블과 상품 ↔ 옵션 간 연관 관계를 맺어주는 product_options 테이블을 생성하여 상품 옵션을 구현했습니다.
요구 사항을 읽어보니 상품 개수처럼 product_options 자체에도 값이 하나 존재해야 했기 때문에 `@ManyToMany` 어노테이션을 사용하지 않고 하나의 Entity로 생성 한 뒤 `@ManyToOne`, `@OneToMany`를 사용하여 연관관계를 설정 해주었습니다. 

Entity에 `@NotNull`, `@Size` 와 같은 어노테이션을 사용하여 요구사항을 만족하였고 ProductOptions에 UNIQUE 설정을 하여 하나의 제품에 같은 상품 옵션이 붙지 않도록 구현하였습니다.

테스트 코드도 작성해보았으나 아직 익숙하지 않아 미흡 할 수 있습니다! 일단 4개 다 통과는 했습니다.

sweet-mine committed:
options, product_options 테이블 정의 및 엔티티 매핑

sweet-mine committed:
Option 엔티티에 value 필드 추가

sweet-mine committed:
Option 엔티티의 value 필드 ProductOption으로 이동

sweet-mine committed:
테스트 케이스 작성 및 디버깅

sweet-mine committed:
README.md 최신화

urarik commented on src/main/java/gift/entity/Option.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.List;
+
+@Entity
+@Table(name = "options")
```
테이블 이름에 복수형을 쓴 이유가 있을까요~?

sweet-mine replied:
옵션 여러개가 들어간다 생각하여 복수형으로 작성했습니다. 구글링해보니 테이블명은 단수형이 권장된다는 것을 배웠으며 수정 하겠습니다!

urarik commented on src/main/java/gift/entity/ProductOption.java:
```diff
@@ -0,0 +1,39 @@
+    @JoinColumn(name = "option_id")
+    private Option option;
+
+    private Long optionValue;
```
optionValue라는 네이밍이 해당 필드를 잘 설명해줄 수 있을까요?

sweet-mine replied:
처음에는 value라는 네이밍을 사용하려 했으나 데이터베이스의 예약어와 겹쳐 에러가 발생했습니다. 그래서 임시로 optionValue라는 네이밍을 사용했으며 좀 더 나은 네이밍을 생각해보고 수정해보겠습니다!

urarik replied:
@sweet-mine 제 생각엔 value라는 단어가 너무 일반적이라 의미 표현을 잘 못해주는 것 같아요~

urarik commented on src/main/java/gift/entity/ProductOption.java:
```diff
@@ -0,0 +1,39 @@
+                columnNames = {"product_id", "option_id"}
+        )
+)
+public class ProductOption {
```
코드와 적어주신 말씀을 보니 Product와 Option이 N:N 관계이고 그렇기 때문에 이어주는 테이블을 하나 만들어준 것 같은데요~ N:N이라면 하나의 Option이 여러 Product와도 연결이 될 수 있어야 하는데 그런 케이스가 있을까요?
만약 1:N 관계라면 불필요하게 테이블이 하나 더 추가된 것 아닐까요~?

sweet-mine replied:
할인율, 태그 같은 상품 옵션은 여러 개의 상품이 해당 옵션을 공유 할 수 있다 생각했고 추후에 옵션을 통해서 상품을 찾는 기능도 구현 가능 할 것 같아 N:N 관계로 설계했습니다. (예시 : '할인율' 옵션을 가진 상품 리스트를 조회하여 할인 상품만 찾기)

urarik commented on src/main/java/gift/repository/ProductOptionRepository.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.data.repository.CrudRepository;
+
+
+public interface ProductOptionRepository extends CrudRepository<ProductOption, Long> {
```
OptionRepository는 JpaRepository를 상속했고 ProductOptionRepository는 CrudRepository를 상속했는데 둘의 차이가 뭘까요~?

sweet-mine replied:
CrudRepository는 이름 그대로 단순한 CRUD 기능만 가진 인터페이스이고 JpaRepository는 거기에 추가로 페이징 기능, 영속성 컨택스트 관리 등이 포함된 인터페이스입니다. 인텔리제이의 자동완성기능을 쓰다 실수로 CrudRepository를 집어넣은 것 같습니다... 수정하겠습니다

urarik commented on src/main/java/gift/repository/ProductOptionRepository.java:
```diff
@@ -0,0 +1,9 @@
+
+
+public interface ProductOptionRepository extends CrudRepository<ProductOption, Long> {
+    ProductOption saveAndFlush(ProductOption productOption);
```
flush는 테스트 코드에서 사용하기 위해 하신건가요? 만약 그렇다면 다른 방법으로 해결해보는 것이 좋아요~
테스트를 위해 작성한 코드가 비즈니스 코드에 들어가면 유지보수가 어려워집니다!

sweet-mine replied:
아하.... 알겠습니다 테스트 코드는 오로지 테스트 코드 내에서만 해결하도록 수정해보겠습니다.

urarik reviewed (changes requested):
수고많으셨습니다~ 코멘트 몇 개 드렸으니 확인해보세요! 그리고 아직 Option을 추가하는 API가 없어보입니다. 추가해보시면 좋을 것 같아요~

sweet-mine committed:
코드 피드백 반영

1. 테이블명 단수형으로 수정

2. product_option 테이블의 option_value를 value로 네이밍 변경

3. ProductOptionRepository가 상속받는 인터페이스를 CrudRepository에서 JpaRepository로 변경

4. 비즈니스 코드에 영향을 주던 테스트 코드 수정

5. Option 관련 API 추가

sweet-mine commented:
코드 피드백 반영했습니다!

1. 테이블명 단수형으로 수정

2. product_option 테이블의 option_value를 value로 네이밍 변경

3. ProductOptionRepository가 상속받는 인터페이스를 CrudRepository에서 JpaRepository로 변경

4. 비즈니스 코드에 영향을 주던 테스트 코드 수정

5. Option 관련 API 추가

urarik reviewed (approved):
한 주동안 수고 많으셨습니다! 머지할게요~
